File: src/processing.py
from re import fullmatch

import unicodedata


def is_match(word, dic):
    if len(word) >= 7:
        return word[3] in dic and word[5] in dic and word[6] in dic
    return False


def has_seven_alphabet_chars(word):
    return fullmatch('[A-zÀ-ú]{7}[/]*', word)


def clean_word(word):
    return word.split('/', 1)[0]


def clean_and_normalize_word(word):
    return unicodedata.normalize('NFKD', word.split('/', 1)[0]).encode('ASCII', 'ignore').decode('ASCII').upper()


def convert_to_leet(word, dic):
    leet = list(word)
    n = [3, 5, 6]
    for i in n:
        leet[i] = dic[word[i]]

    return ''.join(leet)

File: js/processing.js
const ASSOCIATIONS = {
  'A': '4',
  'B': '8',
  'E': '3',
  'G': '6',
  'I': '1',
  'O': '0',
  'Q': '9',
  'S': '5',
  'T': '7',
  'Z': '2'
};

function cleanWord(word) {
  return word.split('/')[0].trim();
}

function normalizeWord(word) {
  return word.normalize('NFKD').replace(/[^\w]/g, '');
}


function hasSevenAlphabetChars(word) {
  return word.match(/^[A-zÀ-ú]{7}$/) !== null;
}

function isMatch(word) {
  if (word.length < 7) {
    return false;
  }

  return normalizeWord(word[3]).toUpperCase() in ASSOCIATIONS &&
    normalizeWord(word[5]).toUpperCase() in ASSOCIATIONS &&
    normalizeWord(word[6]).toUpperCase() in ASSOCIATIONS;
}


function convertToLeet(word) {
  let leet = word.split('').map((char) => char.toUpperCase());

  for (let i of [3, 5, 6]) {
    leet[i] = ASSOCIATIONS[leet[i]];
  }

  return leet.join('');
}

module.exports = {
  normalizeWord,
  isMatch,
  hasSevenAlphabetChars,
  cleanWord,
  convertToLeet,
}